### DISCUSSION TITLE
when use 7z, i found the error like this"Wrong update mode"

### DISCUSSION THREAD

**Discussion by flutelifh1980** [rikyoz/bit7z#211] [Q&A] [answered]
my code is c++, like this: std::wstring zipPath = FilePath::Combine(targetPath, targetZipName) + L".7z";
bit7z::Bit7zLibrary lib("7za.dll");
BitFileCompressor compressor{ lib, BitFormat::SevenZip };
compressor.setCompressionLevel(BitCompressionLevel::Fast);
compressor.compressDirectory(utils::toString(targetFolderPath), utils::toString(zipPath));

i run it, finally , the error is " failed to zip cache dir. error info: Cannot update the existing archive: Wrong update mode." . my bit7z version is 23.01.

**Comment by rikyoz** [comment, marked as answer]
Hi!
The error is due to the fact that the output archive file `zipPath` already exists on the filesystem.
By default, bit7z throws an error in such case.

If you want to update the existing archive, you need to set the correct `UpdateMode`:
```cpp
compressor.setUpdateMode(UpdateMode::Append); // New items will be appended to the existing archive.
// or
compressor.setUpdateMode(UpdateMode::Update); // New items whose path already exists in the archive will overwrite the old ones, other will be appended.
```

If instead you want to entirely overwrite the existing archive file, you need to set `OverwriteMode` as follows:
```cpp
compressor.setOverwriteMode(OverwriteMode::Overwrite);
```
In this case, the existing archive will be replaced with the new one.

I just realized that the error message "Wrong update mode" is not exactly clear on which is the issue. I'll probably improve it in a next version of the library.

**Comment by rikyoz** [comment]
As an aside, I've seen you use `std::wstring` for strings like `zipPath`.
If you are only targeting Windows, you might consider building bit7z with the `-DBIT7Z_USE_NATIVE_STRING=ON` option: in this case bit7z will use `std::wstring`s for all its string function arguments.
This allows you to avoid using `utils::toString` for `std::wstring` -> `std::string` conversions.

**Comment by flutelifh1980** [comment]
ok, thanks. i try to test this according to your method. thanks.
